#18: Implementation of variable $started-volume-first-content-page

PaulRambags opened the pull request:
Solves #18.

@kalaspuffar @bertfrees Please review this pull request. Many thanks!

The build fails because the API was changed.

PaulRambags committed:
#18: Implementation of variable $started-volume-first-content-page

PaulRambags committed:
#18: Removed metaVolumeFirstContentPage everywhere in the code and modifications were made so that variable started-volume-first-content-page is available only in the context of a toc-entry-on-resumed element.

PaulRambags commented:
@bertfrees Please check this pull request. Thank you!

PaulRambags committed:
#18: Initialisation of the meta variables fixed

bertfrees commented on src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java:
```diff
@@ -132,12 +132,19 @@ public void endEntry() {
      * and margins that should be kept. That said, the previous implementation did not handle
      * borders and margins correctly either, so fixing this issue can be seen as an optimization.</p>
+     * 
+     * <p>This method sets the meta page number of the range blocks that are filtered out.</p> 
```
"sets the meta page number of ... _to the provided value_." ?

bertfrees replied:
@PaulRambags Have you seen above comment? Minor thing, just so that there can not be confusion about who provides the values (i.e. the caller, not the function).

PaulRambags replied:
Done.

bertfrees reviewed (commented):
Your `MetaVariable` class is in se a good addition. What I like about it is that this way you make the possible meanings of `Context.getMetaPage()` explicit.

However the issue is that to me, by passing around the `MetaVariable` objects in the `obfl` package, you create the impression that the OBFL parser is determining the meaning of the field, while in fact this happens in the formatter. At least it can create that confusion. The formatter uses the field for different purposes in different contexts, and the OBFL layer is only responsible for binding the right variable name to the field. This is how we have defined `metaPage` in the API: as a general-purpose field (https://github.com/mtmse/dotify.api/pull/10). Maybe this was not the best choice after all, maybe your `MetaVariable` class should be part of the API? I'm not sure anymore. The way we have done it now is also a valid approach I think.

Anyway, I guess what I'm trying to say is just that we have to make the intend very clear with documentation to avoid misunderstandings. I'm not totally happy with the current Javadoc of `MetaVariable`. I'll make a change suggestion next week.

Overall very good change though!

bertfrees committed:
Rework OBFLExpression code a bit

- Generalize the "MetaVariable" enum to "OBFLVariable" to include all
  variables, not only the "meta" ones.
- Which variables are enabled in an OBFL expression is now made
  explicit in ObflParserImpl.
- Setting non-default variable names is now done with a single
  setVariableName method with OBFLVariable argument.
- Improve documentation

bertfrees commented:
@PaulRambags This is what I made of it: https://github.com/snaekobbi/dotify.formatter.impl/commit/847a7cf47f99dd17428869196afe9b73a65523a2

PaulRambags commented on src/org/daisy/dotify/formatter/impl/obfl/OBFLExpressionBase.java:
```diff
@@ -4,118 +4,161 @@
+                throw new IllegalArgumentException(); // coding error
+            }
+            this.variables.add(v);
+            setVariableName(v, null);
```
@bertfrees my IDE (NetBeans) warns me against an "Overridable method call in constructor".
I will fix this.

For more details, please see: https://stackoverflow.com/questions/3404301/whats-wrong-with-overridable-method-calls-in-constructors

bertfrees replied:
RIght good point.

PaulRambags committed:
#18: Fixed overridable method call in constructor of OBFLExpressionBase

PaulRambags commented:
@bertfrees Please review and approve this pull request, thank you

bertfrees commented on src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:
```diff
@@ -1641,8 +1641,6 @@ private void parseVolumeTemplate(
             OBFLVariable.SHEET_COUNT,
             OBFLVariable.VOLUME_SHEET_COUNT);
-        condition.setVariableName(OBFLVariable.VOLUME_COUNT, getAttr(event, "volume-count-variable"));
-        condition.setVariableName(OBFLVariable.VOLUME_NUMBER, getAttr(event, "volume-number-variable"));
```
@PaulRambags Isn't this a regression?

bertfrees replied:
Hmm I see you created the issue https://github.com/mtmse/obfl/issues/13. Fine by me. But maybe a warning should be created when the attribute is specified. By the way there is another comment (a FIXME) somewhere in this file about it.

PaulRambags replied:
Good point. I'll throw an `UnsupportedOperationException`. This exception could be removed from the code in a later stadium.

bertfrees replied:
Great. Can you also add the exception [here](https://github.com/mtmse/dotify.formatter.impl/blob/ba9320a09f95ecc67741f1a4e116c576c585320f/src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java#L267-L268)?

https://github.com/mtmse/dotify.formatter.impl/blob/ba9320a09f95ecc67741f1a4e116c576c585320f/src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java#L267-L268

Actually I have been using this attribute in Pipeline, being totally unaware it is ignored. Luckily I just set it to the default value "page" :).

PaulRambags replied:
>Actually I have been using this attribute in Pipeline, being totally unaware it is ignored.

Beware! In the coming release, an exception will be thrown even if you use the default value. ;-)

bertfrees replied:
Yes, sure, I'm removing it.

PaulRambags committed:
#18: Issue an UnsupportedOperationException when an alternative variable name is used

PaulRambags committed:
#18: An UnsupportedOperationException is thrown when "page-number-variable" is used. As a result, quite a lot of OBFL input test files had to be modified.

PaulRambags commented:
@kalaspuffar The PR now includes a complete abandoning of alternative variable names. Only the default variable names are supported. This may have an impact on Dotify Studio - please check this out. In particular, it occured to me that (almost?) all OBFL input test files have something like this:
```
<layout-master name="..." page-width="..." page-height="..." duplex="true" page-number-variable="page">
```
That must be changed to:
```
<layout-master name="..." page-width="..." page-height="..." duplex="true">
```

bertfrees commented:
Most of those tests came from me ("dp2" = "DAISY Pipeline 2").

kalaspuffar commented:
This PR is now approved, I will run regression tests.
